docs/dev/design/file-tree-diff.rst

```diff
@@ -0,0 +1,234 @@
+The result is easy to parse.
+
+To start, we will only consider HTML files (``--include=*.html``).
+
+Lines changed between two files
+-------------------------------
+
+Once we have the list of files that changed, we can use a tool like ``diff`` to get the lines that changed.
+This is useful to link to the most relevant files that changed in a PR.
+
+.. code:: console
+
+   $ cat a.txt
+   One
+   Two
+   Three
+   Four
+   Five
+   $ cat b.txt
+   Ore
+   Three
+   Four
+   Five
+   Six
+   $ diff --side-by-side --suppress-common-lines a.txt b.txt
+   One                                                           | Ore
+   Two                                                           <
+                                                                 > Six
+
+.. note::
+
+   Taken from https://stackoverflow.com/questions/27236891/diff-command-to-get-number-of-different-lines-only.
+
+The command will return only the lines that changed between the two files.
+We can just count the lines, or maybe even parse each symbol to check if the line was added or removed.
+
+Another alternative is to use the `difflib <https://docs.python.org/3/library/difflib.html>`__ module,
+the only downside is that it doesn't distinguish lines that were changed from lines that were added or removed.
+But maybe that's ok? Do we really need to know if a line was changed instead of added or removed?
+
+.. code:: python
+
+   import difflib
+
+   diff = difflib.ndiff(["one", "two", "three", "four"], ["ore", "three", "four", "five"])
+   print(list(diff))
+   # ['+ ore', '- one', '- two', '  three', '  four', '+ five']
+
+A good thing of using Python is that we don't need to write the files to disk,
+and the result is easier to parse.
+
+Storing results
+---------------
+
+Doing a diff between two versions can be expensive, so we need to store the results.
+
+We can store the results in the DB (``VersionDiff``), or in S3 (``diff/project/version-a/version-b.json``), or maybe a combination of both.
+The information to store would contain some information about the versions compared, the build, and the diff itself.
+
+.. code:: json
+
+   {
+       "version_a": {"id": 1, "build": {"id": 1}},
+       "version_b": {"id": 2, "build": {"id": 2}},
+       "diff": {
+           "added": [{"file": "new.txt"}],
+           "removed": [{"file": "deleted.txt"}],
+           "modified": [{"file": "changed.txt", "lines": {"added": 1, "removed": 1}}]
+       }
+   }
+
+The information is stored in a similar way that it will be returned by the API.
+Things important to note:
+
+- We need to take into consideration the diff of the latest successful builds only.
+  If any of the builds from the stored diff don't match the latest successful build of any of the versions,
+  we need to the diff again.
+- Once we have the diff between versions ``A`` and ``B``, we can infer the diff between ``B`` and ``A``.
+  We can store that information as well, or just calculate it on the fly.
```

Comment:
It's useful to see the diff between other versions, stable vs latest, v1 vs v2, etc. This feature isn't just useful for PR builds.

Reply:
Can you expand on what are those use cases you have in mind to see the files differences between those other versions?

Reply:
The sames as checking the diff between a PR preview and latest/stable. You want to know which files has changed, suggestions for possible redirects.

Reply:
You can see a case where a v1 is out and a v2 is released, and you need to figure out what files are different to create redirects on `latest`. So you'd want a diff between `latest` and `v2`, or similar. Or even a diff between `v1` and `v2` to better understand the work done. 

It doesn't seem like a super common case though, but I can see some uses for it. It seems safe to avoid doing that work in the initial version of this, and see if people request it.

Reply:
> It doesn't seem like a super common case though, but I can see some uses for it. It seems safe to avoid doing that work in the initial version of this, and see if people request it.

Restricting the versions that are allowed would add a special case to the API, I mean, is not a tone of work, but allowing all versions isn't either.

If we are going to restrict the versions from `version_a` to external versions, are we going to restrict the versions from `version_b`? Just latest or stable?

Reply:
The API could always return 404 or similar HTTP error code if there is no PR involved in the comparison; for example. I think that's not a problem and should be easy to implement.

I'd start with `version_a` being the `default version` or `latest` and `version_b` being an external version.